lib/storage/WebStorage.js

```diff
@@ -27,10 +32,23 @@ const webStorage = {
             .then(() => raiseStorageSyncEvent(key));
 
         // If we just call Storage.clear other tabs will have no idea which keys were available previously
-        // so that they can call keysChanged for them. That's why we iterate and remove keys one by one
-        this.clear = () => Storage.getAllKeys()
-            .then(keys => _.map(keys, key => this.removeItem(key)))
-            .then(tasks => Promise.all(tasks));
+        // so that they can call keysChanged for them. That's why we iterate over every key and raise a storage sync
+        // event for each one
+        this.clear = () => {
+            let allKeys;
+
+            // They keys must be retreived before storage is cleared or else the list of keys would be empty
+            return Storage.getAllKeys()
+                .then((keys) => {
+                    allKeys = keys;
+                })
+                .then(() => Storage.clear())
+                .then(() => {
+                    // Now that storage is cleared, the storage sync event can happen so that it is more of an atomic
+                    // action
+                    _.each(allKeys, raiseStorageSyncEvent);
+                });
+        };
```

Comment:
You can avoid the closure variable by doing something like

```js
        this.clear = () => {
            return Storage.getAllKeys()
                .then((allKeys) => {
                    return Storage.clear()
                        .then(() => _.each(allKeys, raiseStorageSyncEvent))
                });
        };
```

Although code has more nesting, it conveys the information you have to explain using comments here 

I guess this is opinion based and there's no hard rule on how to style such concepts

Reply:
Makes sense, yeah. I prefer the clarity of my current code than trying to reason about those nested promises. Thanks!